Add zooming feature.

carloslfu opened the pull request:
Zoom in, out and reset with buttons and mouse scrollwheel.

![image](https://cloud.githubusercontent.com/assets/5993168/8608966/ca5a4366-2665-11e5-941c-f345ae57bb99.png)

![image](https://cloud.githubusercontent.com/assets/5993168/8609046/ab487f46-2666-11e5-82f4-f4b7bc61cef3.png)

alloplastic commented:
Hi, Carlos (@carloslfu).

This is great... I pulled down the PR and got it running.  The buttons are a nice touch.  Within about 2-3 weeks I should be actively integrating blocks into the app I'm building.  I'll use this PR as the starting point.

Thanks for circling back to this.  I got distracted myself just getting a basic app layout working.

Sheldon

carlosperate commented on core/inject.js:
```diff
@@ -416,6 +429,12 @@ Blockly.init_ = function(mainWorkspace) {
+  //mouse-wheel for other browsers
+  Blockly.bindEvent_(svg, 'mousewheel', null, Blockly.onMouseWheel_);
+  //mouse-move for tracking mouse position
+  Blockly.bindEvent_(svg, 'mousemove', null, Blockly.onMouseMoveTracking_);
```
These three should probably only be bound if the zoom feature is selected.

carlosperate commented on core/workspace_svg.js:
```diff
@@ -91,6 +92,42 @@ Blockly.WorkspaceSvg.prototype.scrollX = 0;
+ * Current scale speed.
+ * @type {number}
+ */
+Blockly.Workspace.prototype.scaleSpeed = 1.2;
```
It's hard to understand what this number indicates without some kind of unit of measurement. I'd be useful to have that in this jsdoc.

carlosperate commented on core/block_svg.js:
```diff
@@ -1035,13 +1038,12 @@ Blockly.BlockSvg.prototype.disposeUiEffect = function() {
   // Start the animation.
   clone.startDate_ = new Date();
-  Blockly.BlockSvg.disposeUiStep_(clone, this.RTL);
+  Blockly.BlockSvg.disposeUiStep_.bind(this)(clone, this.RTL);
```
Javascript is definitely not my main language, but is this correct syntax? what is the second parenthesis doing?

NeilFraser replied:
Indeed, I'd just say:
Blockly.BlockSvg.disposeUiStep_.call(this, clone, this.RTL);
No need to create a new function object with each call.

carlosperate commented on core/block_svg.js:
```diff
@@ -1094,17 +1097,18 @@ Blockly.BlockSvg.prototype.connectionUiEffect = function() {
-      Blockly.BlockSvg.connectionUiStep_(ripple);
-    };
+      this.connectionUiStep_(ripple);
+    }.bind(this);
```
??

NeilFraser replied:
Why not use the currently unused thisBlock_ variable instead of bind().

carlosperate commented on core/blockly.js:
```diff
@@ -514,23 +544,22 @@ Blockly.getMainWorkspaceMetrics_ = function() {
+                             contentX + viewWidth);
+    var topEdge = Math.min(contentY - viewHeight * this.scale / 2,
+                           contentY + contentHeight - viewHeight);
+    var bottomEdge = Math.max(contentY + contentHeight + viewHeight * this.scale / 2,
```
This line is over 80 characters.

carlosperate commented on core/comment.js:
```diff
@@ -94,6 +94,10 @@ Blockly.Comment.prototype.createEditor_ = function() {
+  //dont zooming with mousewheel
+  Blockly.bindEvent_(this.textarea_, 'mousewheel', this, function (e){e.stopPropagation()});
+  //for firefox
+  Blockly.bindEvent_(this.textarea_, 'DOMMouseScroll', this, function (e){e.stopPropagation()});
```
Both over 80 characters.

carlosperate commented on core/flyout.js:
```diff
@@ -619,8 +621,32 @@ Blockly.Flyout.prototype.createBlockFunc_ = function(originBlock) {
+      xyOld.x /= Blockly.mainWorkspace.scale;
+      xyOld.y /= Blockly.mainWorkspace.scale;
+      //Calculate the position to create the block, fixing scale
+      var xyCanvastoSvg = Blockly.getRelativeXY_(Blockly.mainWorkspace.getCanvas());
```
Over 80 characters.

takaokouji commented:
Awesome! :+1:  I want to merge this feature master.

carlosperate commented on core/utils.js:
```diff
@@ -280,22 +280,58 @@ Blockly.getRelativeXY_ = function(element) {
+      canvasFlag = true;
+    }
+    //before the svg canvas scale the coordinates
+    if(canvasFlag) {
```
space between `if` and parenthesis.

carlosperate commented on core/utils.js:
```diff
@@ -304,14 +340,42 @@ Blockly.getSvgXY_ = function(element) {
+  e.getBBox = function() {
+    //fix scale if the element is a svg canvas or a child
+    var BBox = getBBox.call(e);
+    if(workspace) {
```
parenthesis space

carlosperate commented on core/workspace_svg.js:
```diff
@@ -681,6 +752,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
+ * @param {!number} Y coordinate of center.
+ * @param {!number} type Type of zomming (-1 zooming out and 1 zooming in).
+ */
+Blockly.WorkspaceSvg.prototype.zoom  = function(x ,y , type) {
```
space before comma instead of after

carlosperate commented on core/workspace_svg.js:
```diff
@@ -681,6 +752,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
+/**
+ * Zooming the blocks centered in (x,y) coordinate with zooming in or out.
+ * @param {!number} x X coordinate of center.
+ * @param {!number} Y coordinate of center.
```
lower case y variable name

carlosperate commented:
Hey Carlos, 

Great work again! I haven't had a chance to test it yet, I'll probably add more comments once I get to check the code live, but I've got a couple of questions so far:

* It seems like the mouse position is only required on scroll or block creation on flyout, if that is the case, wouldn't it be better for performance to check the mouse position only when strictly necessary in those functions rather than have a event constantly firing on mouse movement? 

* This feature seems to have the scroll wheel automatically bind to the zoom if turned on, could we have a scroll wheel flag? It could still be set to `true` by default, but implementations with limited vertical space might appreciate to still be able to scroll.

* Will this work with multiple workspaces? I didn't see many arguments for a specific workspace (although the ones present might be the ones that matter for this feature) and mostly calls to `getMainWorkspace`.

Also, this is going to sound really picky, and it might be (including some of my comments in the code already), but it'd be good to stick to the style indicated in http://google.github.io/styleguide/javascriptguide.xml, which defers to the c++ guide for general white-spacing http://google.github.io/styleguide/cppguide#Horizontal_Whitespace

Yeah, great work! I'll give it a spin once I get a chance.

alloplastic commented:
Hi, Carlos.

I wanted to let you know about a couple of errors I got while integrating the zoom feature into my app.  This may have something to do with the organization of my codebase... but I'll list the changes I had to make in case they are relevant.

The issue was that three references to "workspace" were failing in workspace_svg.js.   Replacing them with "Blockly.mainWorkspace" fixed the problem.  

*In Blockly.WorkspaceSvg.prototype.zoom()*

     center = center.matrixTransform(workspace.getCanvas().getCTM().inverse());
       ...
      var canvas = workspace.getCanvas();


*In Blockly.WorkspaceSvg.prototype.zoomReset()*

     workspace.scrollbar.set(-metrics.contentLeft, -metrics.contentTop);
 
Cheers,
Sheldon

carlosperate commented:
WorkspaceSvg extendex Workspace, so a reference to `this` works then it would probably be better, in case we are trying to programatically change the zoom level of a workspace not on focus.

alloplastic commented:
Ah, that makes sense.  I made the change, and that works just fine in my build, too.

Thanks,
Sheldon

carloslfu commented:
Hi Carlos (@carlosperate),

Thanks for your comments, I'll make the appropriate changes and fixes as soon as possible. I removed mouse tracking because it is unnecessary, I have added the scrollWheel flag to inject options (default to true) and some fixes.

cleaverpj commented:
This is awesome, nice work.  How will I incorporate it into my app?

carloslfu commented:
There are some things that need to be resolved but you can merge in this branch in the master branch, [build blockly](https://developers.google.com/blockly/hacking/building) and use in the compressed form.

NeilFraser commented on core/inject.js:
```diff
@@ -455,6 +480,20 @@ Blockly.init_ = function(mainWorkspace) {
     mainWorkspace.scrollbar = new Blockly.ScrollbarPair(mainWorkspace);
     mainWorkspace.scrollbar.resize();
   }
+  if (options.enableZoom != undefined) {
```
!==

NeilFraser commented on core/blockly.js:
```diff
@@ -322,6 +322,24 @@ Blockly.onMouseMove_ = function(e) {
+    var e = window.event || e; // old IE support
+    var delta = Math.max(-1, Math.min(1, (e.wheelDelta || -e.detail)));
+    var position = Blockly.mouseToSvg(e, workspace.options.svg);
+    workspace.zoom (position.x , position.y, delta);
```
Remove extra space before parens.

NeilFraser commented on core/comment.js:
```diff
@@ -94,6 +94,14 @@ Blockly.Comment.prototype.createEditor_ = function() {
   this.foreignObject_.appendChild(body);
   Blockly.bindEvent_(this.textarea_, 'mouseup', this, this.textareaFocus_);
+  //dont zooming with mousewheel
+  Blockly.bindEvent_(this.textarea_, 'mousewheel', this, function (e){
```
Spacing change:  "function(e) {"
And line 102.

NeilFraser commented on core/flyout.js:
```diff
@@ -620,8 +622,35 @@ Blockly.Flyout.prototype.createBlockFunc_ = function(originBlock) {
-    var xyNew = Blockly.getSvgXY_(svgRootNew);
-    block.moveBy(xyOld.x - xyNew.x, xyOld.y - xyNew.y);
+    //if flyout is inside of canvas, fix scale
+    if (flyout.targetWorkspace_ === Blockly.mainWorkspace) {
```
Don't need === in this case (== is fine).

NeilFraser commented on core/flyout.js:
```diff
@@ -657,11 +686,17 @@ Blockly.Flyout.prototype.getRect = function() {
   var BIG_NUM = 10000000;
   var x = Blockly.getSvgXY_(this.svgGroup_).x;
-  if (!this.RTL) {
+  if (!Blockly.RTL) {
```
Blockly has no RTL property.  Workspaces have RTL (and potentially different workspaces might have different settings from each other).

NeilFraser commented on core/flyout.js:
```diff
@@ -657,11 +686,17 @@ Blockly.Flyout.prototype.getRect = function() {
-      BIG_NUM + this.width_, this.height_ + 2 * BIG_NUM);
+  //fix scale if is descendant of bubble canvas
+  if (Blockly.isDescendant_(this.svgGroup_, Blockly.mainWorkspace.getBubbleCanvas())) {
+      return new goog.math.Rect(x, -BIG_NUM,
```
Indents should be 2 spaces.

NeilFraser commented on core/utils.js:
```diff
@@ -281,22 +281,58 @@ Blockly.getRelativeXY_ = function(element) {
+ * Return true if child is descendant of parent element.
+ * @param {!Element} child Reference element.
+ * @param {!Element} parent Element to search in the parents of child.
+ * @return {!Boolean} Is descendant or not.
```
{boolean} since we want the primitive (true) not a Boolean object (new Boolean(true)).

NeilFraser commented on core/utils.js:
```diff
@@ -281,22 +281,58 @@ Blockly.getRelativeXY_ = function(element) {
+ * @private
+ */
+Blockly.isDescendant_ = function(child, parent) {
+     var node = child.parentNode;
```
2 space indents.

NeilFraser commented on core/utils.js:
```diff
@@ -281,22 +281,58 @@ Blockly.getRelativeXY_ = function(element) {
+ */
+Blockly.isDescendant_ = function(child, parent) {
+     var node = child.parentNode;
+     while (node != null) {
```
while (node) {

NeilFraser commented on core/utils.js:
```diff
@@ -281,22 +281,58 @@ Blockly.getRelativeXY_ = function(element) {
+ * @return {!Boolean} Is descendant or not.
+ * @private
+ */
+Blockly.isDescendant_ = function(child, parent) {
```
Why not just use goog.dom.contains(parent, descendant)?
https://closure-library.googlecode.com/git-history/docs/local_closure_goog_dom_dom.js.source.html#line1312

NeilFraser commented on core/utils.js:
```diff
@@ -305,14 +341,42 @@ Blockly.getSvgXY_ = function(element) {
  * @param {string} name Element's tag name.
  * @param {!Object} attrs Dictionary of attribute names and values.
  * @param {Element=} opt_parent Optional parent on which to append the element.
+ * @param {Workspace=} workspace Optional workspace for access to context (scale...).
```
Optional arguments must be named opt_workspace.
Also, the type would be {Blockly.Workspace=}

NeilFraser commented on core/utils.js:
```diff
@@ -305,14 +341,42 @@ Blockly.getSvgXY_ = function(element) {
+      do {
+        // Loop through this block and every parent.
+        var xy = Blockly.getRelativeXY_(element);
+        if (element === workspace.getCanvas()) {
```
== works fine here.

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -91,6 +92,42 @@ Blockly.WorkspaceSvg.prototype.scrollX = 0;
+
+/**
+ * Current position of mouse.
+ * @type {number}
```
{?number} since it is nullable.

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -732,6 +803,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
 
+/**
+ * Zooming the blocks centered in (x, y) coordinate with zooming in or out.
+ * @param {!number} x X coordinate of center.
```
Primitives are not nullable by default, so just {number}.

To clarify in general:
{Foo} -> Foo object or null
{!Foo} -> Foo object
{string} -> string primitive (not a String)
{?string} -> string primitive or null

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -732,6 +803,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
+  var speed = this.scaleSpeed;
+  var metrics = this.getMetrics();
+  var center = this.options.svg.createSVGPoint();
+  var g = this.getCanvas();
```
Unused variable.

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -732,6 +803,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
+  var y = center.y;
+  var canvas = workspace.getCanvas();
+  // scale factor
+  var scale = (type == 1)?speed:1/speed;
```
Add 6 spaces around 3 operators.

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -732,6 +803,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
+ */
+Blockly.WorkspaceSvg.prototype.zoomCenter  = function(type) {
+  var metrics = this.getMetrics();
+  var x = metrics.viewWidth/2;
```
More spaces.  And next line.

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -732,6 +803,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
+ * Updates the grid pattern.
+ */
+Blockly.WorkspaceSvg.prototype.updateGridPattern_  = function() {
+  this.options.gridPattern.setAttribute('width', this.options.gridOptions.spacing * this.scale+'');
```
Setting an attribute does not need string conversion.  The browser will handle it.

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -732,6 +803,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
+Blockly.WorkspaceSvg.prototype.updateGridPattern_  = function() {
+  this.options.gridPattern.setAttribute('width', this.options.gridOptions.spacing * this.scale+'');
+  this.options.gridPattern.setAttribute('height', this.options.gridOptions.spacing * this.scale+'');
+  var half = Math.floor(this.options.gridOptions['spacing'] / 2) + .5;
```
0.5

NeilFraser commented on core/workspace_svg.js:
```diff
@@ -732,6 +803,87 @@ Blockly.WorkspaceSvg.prototype.markFocused = function() {
+  half *= this.scale;
+  start *= this.scale;
+  end *= this.scale;
+  if (this.options.gridPattern.children[0]) {
```
Can we please save this.options.gridPattern.children[0] and this.options.gridPattern.children[1] out to variables?

NeilFraser commented on core/zoom-controls.js:
```diff
@@ -0,0 +1,245 @@
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2011 Google Inc.
```
2015

NeilFraser commented on core/zoom-controls.js:
```diff
@@ -0,0 +1,245 @@
+ * @type {string}
+ * @private
+ */
+Blockly.ZoomControls.prototype.ZOOMIN_URL_ = 'zoomin.svg';
```
// TODO: Add these images to sprites.png so that we don't have three separate http calls.

This can be done later, separate images are fine for now.

NeilFraser commented on core/zoom-controls.js:
```diff
@@ -0,0 +1,264 @@
+ * @type {string}
+ * @private
+ */
+Blockly.Trashcan.prototype.SPRITE_URL_ = 'media/sprites.png';
```
Trashcan?

carloslfu replied:
Ups!

NeilFraser commented:
I've sent a PR that makes a few changes:
* zoom-controls.js renamed to zoom_controls.js (to match all other files except a 3rd party library for realtime).
* Default zooming to off, instead of on.  The default config should result in a minimal Blockly.
* Refactor zoom config to use the existing option object already on workspaces.
* Some comment editing.

A couple of remaining issues I've noticed are:
* Mouse wheel zooming while dragging results in a delta between the mouse pointer and the block.
* Editable text input fields don't scale at all.

NeilFraser commented:
(Don't worry about the branch conflicts with master, it's just the compiled JS.)

I'd like to take a step back for a moment and talk about the purpose of zooming.  There are two that come to mind.
1) Zooming in for easier use (e.g. fat fingers on high resolution screens, such as phones).
2) Zooming out for a high level view of larger programs.
Before we go further, are these the only use cases?  Are we missing anything?

If those are the only two use cases, then zooming in and zooming out serve quite different purposes from each other and thus should behave differently.
1) Zooming in should behave exactly as this PR behaves.
2) Zooming out should not shrink mutator, comment, or warning bubbles, these should be hidden instead.  Likewise, at some zoom level individual blocks should fuse with their parents to form higher-level groupings (much like collapsed blocks do now).

We don't need to achieve all this before merging with master, but I just want to get a sense of where we are heading with this.

carlosperate commented:
I would also like to argue that on some tablets it'd be useful to zoom out
for normal use, as the browser resolution is usually set pretty low on even
large tablets, so more blocks would fit on the screen and still be usable
with touch.

On Thu, 30 Jul 2015 07:34 Neil Fraser <notifications@github.com> wrote:

> (Don't worry about the branch conflicts with master, it's just the
> compiled JS.)
>
> I'd like to take a step back for a moment and talk about the purpose of
> zooming. There are two that come to mind.
> 1) Zooming in for easier use (e.g. fat fingers on high resolution screens,
> such as phones).
> 2) Zooming out for a high level view of larger programs.
> Before we go further, are these the only use cases? Are we missing
> anything?
>
> If those are the only two use cases, then zooming in and zooming out serve
> quite different purposes from each other and thus should behave differently.
> 1) Zooming in should behave exactly as this PR behaves.
> 2) Zooming out should not shrink mutator, comment, or warning bubbles,
> these should be hidden instead. Likewise, at some zoom level individual
> blocks should fuse with their parents to form higher-level groupings (much
> like collapsed blocks do now).
>
> We don't need to achieve all this before merging with master, but I just
> want to get a sense of where we are heading with this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/blockly/pull/125#issuecomment-126201698>.
>

carloslfu commented:
I find very interesting the higher-level groupings proposed by Neil, this could lead to a new type of blocks that allow to make greater abstractions and handle larger codes. These blocks or groups could only give meaning and provide code abstraction levels. In addition, if activated with the zoom would be amazing.

carloslfu commented:
Taking in mind the issue related to input fields and zoom-out, I propose two different solutions:

* When click into input field automatically zoom-in to center and when blur restore the zoom optionally.
* When click into input field automatically zoom-in only the area of the field like google chrome's auto-zoom feature on mobile devices.

carlosperate commented:
I'm not too sure automatic zooming, I don't see much value added, just a visual effect. I think I would personally find it distracting, specially if it does not come back to the original position on blur. Although, I can definitely see some value if clicking a partially obstructed block (say a block in the workspace that is not fully visible on the screen) shifts the workspace for the block to be fully visible, but not more than necessary, a full shift to the centre can be a bit too much. However, that would not really be directly related to the zooming functionality.

yoderjen commented:
One thing we've done with our project to help improve usability with large programs is that when collapsing a stack of blocks, the top procedure name (if existing) or top comment (if existing) bubble up to become the text displayed on the collapsed block.  This makes it easy to use comments to label code groups, making finding sub-parts of a program much easier and providing more space to the user (our programs tend to be a large number of small-to-medium groups of blocks - this would be less useful if the programs tended to be one enormous chunk of code)

It would be nice if zooming out had a similar effect - if, at a certain level, as you zoom out you could see this text (procedure names and comments have been useful for us) as a way of distinguishing which bit of code each thing is.

carloslfu commented:
Hi carlos (@carlosperate),

Shift obstructed blocks is an useful feature, this could be added in another pull.

Related to text inputs the other option would be implement the text input field in pure SVG avoiding use of the widget-div, the main problem is that SVG foreign text inputs are largely bugged, I'm experimenting with constructing a svg input text field for blockly from scratch in pure SVG. I wonder if anyone has another solution to scale text fields properly.

This is what would be achieved with [pure SVG text inputs](https://carloslfu.github.io/BBlocks.js/demos/basic/demo.html) (you can zoom with mouse scroll wheel in this example), now have many bugs related to events and performance. If pure SVG text inputs are the prefered way this can be done in another pull.

carloslfu commented:
@yoderjen's approach of higer-level groupings looks good.

yoderjen commented:
I had another thought on organization of space for large projects.  It could be nice to have a "mini-map" of the blockly workspace, which would be a small div containing an image of the whole workspace currently being used (not just the visible part, but also the parts with blocks that you've scrolled away from), with a transparent square showing what the current workspace is currently displaying.   clicking on the mini-map would re-center the workspace.  Not sure if this belongs in the zooming discussion, but I thought it might affect implementation so thought I'd throw the idea out there.

NeilFraser commented:
Merged!  But I think I messed up the conflict resolution process in a way that failed to preserve the massive commit history of this PR.  :(

carloslfu commented:
I'm happy to see this feature in the Blockly core :D and I'm eager to collaborate on improving Blockly.
